### PR TITLE
Add support for Scorch, Brittle, and Sap

### PR DESCRIPTION
**Pull request by Quotae**
- Implements Scorch, Brittle, and Sap
- Each ailment becomes enabled globally as long as you have some chance or way to apply them
- Config options for Scorch and Brittle to change their effect like Shock
image

- Outputs in Calcs tab for effect, chance, and duration
image

- Everything that cares about or modifies general "elemental ailments" will interact with the alternate ailments accordingly
image
image

--------------------------

Comments:
- I tested in-game and Scorch, Brittle, and Sap do not activate the area mod on Elemental Hit, so I didn't change it in PoB. I submitted a bug report to GGG on it because it seems like a bug, so I'll make another PR or commit if it gets changed.

- I had to do things kind of strangely due to how it needed to work. The ailments need to be enabled globally if you can apply them, so that something like Elemental Hit has constant access to all three config options, and so you can use them to support your minions or DoT skills(both of which will never be able to apply the ailments). To do this I added two skillFlags, one for when you can inflict an ailment at all, and one for when the skill you've selected can inflict it. With just one flag, the Calc screen outputs would show up regardless of whether the skill actually had the damage/ability to inflict the ailments.

- Related, I also had to add a Condition for when you can inflict each ailment so that I could properly make the effects stop happening if you lose the ability to inflict them. Similar to the CanBeElusive condition so you lose the effects of Elusive if you can't gain it.

- We unfortunately don't know the effect formulas for these alternate ailments like we do for Chill, Freeze, and Shock, so there's no breakdown and calculation there. Can be relatively easily added if those formulas are ever revealed to us.

### CODE REVIEW

**Comment by SabreCat** [comment]
ha, thanks for this!

>I tested in-game and Scorch, Brittle, and Sap do not activate the area mod on Elemental Hit, so I didn't change it in PoB. I submitted a bug report to GGG on it because it seems like a bug, so I'll make another PR or commit if it gets changed.

oh no! my poor jank build! ;_;

File: Modules/CalcOffence-3_0.lua [hunk 1]
```diff
@@ -1645,6 +1667,33 @@ function calcs.offence(env, actor, activeSkill)
 		else
 			output.KnockbackChanceOnHit = skillModList:Sum("BASE", cfg, "EnemyKnockbackChance")
 		end
+		if not skillFlags.hit and skillModList:Sum("BASE", cfg, "ScorchChance") > 0 then
```
Comment (ppoelzl) [inline Modules/CalcOffence-3_0.lua:1670]: Analogue to the above:
```lua
if not skillModList:Sum("BASE", cfg, "ScorchChance") > 0 and not skillFlags.hit then
    output.ScorchChanceOnHit = 0
elseif skillModList:Sum("BASE", cfg, "ScorchChance") > 0 then
    skillFlags.inflictscorch = true
    if skillFlags.hit then
        output.ScorchChanceOnHit = m_min(100, skillModList:Sum("BASE", cfg, "ScorchChance"))
    end
end
```
Repeat for the other two ailments.

File: Modules/CalcSections-3_0.lua [hunk 2]
```diff
@@ -823,11 +823,26 @@ return {
...
+	}, },
+	{ label = "Scorch Dur. Mod", haveOutput = "ScorchDurationMod", flag = "scorch", { format = "x {2:output:ScorchDurationMod}",
+		{ label = "Player modifiers", modName = "EnemyScorchDuration", cfg = "skill" },
+		{ label = "Enemy modifiers", modName = "SelfScorchEffect", enemy = true },
+	}, },
 	{ label = "Inc. Chill Effect", flag = "chill", { format = "{0:output:ChillEffectMod}%",
 		{ label = "Player modifiers", modName = "EnemyChillEffect", cfg = "skill" },
 		{ label = "Enemy modifiers", modName = "SelfChillEffect", enemy = true },
 	}, },
-	{ label = "Chill Dur. Mod:", flag = "chill", { format = "x {2:output:ChillDurationMod}",
```
Comment (ppoelzl) [inline Modules/CalcSections-3_0.lua:830]: For completely unrelated changes like this one, please refactor into another PR.

File: Modules/CalcOffence-3_0.lua [hunk 3]
```diff
@@ -1599,6 +1602,25 @@ function calcs.offence(env, actor, activeSkill)
 		else
 			output.ChillChanceOnCrit = 100
 		end
+		if skillModList:Flag(cfg, "CritAlwaysAltAilments") and skillFlags.hit then
+			skillFlags.inflictscorch = true
```
Comment (ppoelzl) [inline Modules/CalcOffence-3_0.lua:1606]: `inflictscorch` should follow camelCase convention, like so: `inflictScorch`.
Repeat for the other two ailments.

File: Modules/ConfigOptions.lua [hunk 4]
```diff
@@ -888,18 +888,33 @@ return {
 	{ var = "conditionEnemyIgnited", type = "check", label = "Is the enemy Ignited?", implyCond = "Burning", tooltip = "This also implies that the enemy is Burning.", apply = function(val, modList, enemyModList)
 		enemyModList:NewMod("Condition:Ignited", "FLAG", true, "Config", { type = "Condition", var = "Effective" })
 	end },
+	{ var = "conditionEnemyScorched", type = "check", ifFlag = "inflictscorch", label = "Is the enemy Scorched?", tooltip = "Scorched enemies have lowered elemental resistances, up to -30%.\nThis option will also allow you to input the effect of Scorched.", apply = function(val, modList, enemyModList)
```
Comment (ppoelzl) [inline Modules/ConfigOptions.lua:891]: Flags should follow the ones set in `CalcPerform.lua`, in this case `CanInflictScorch`.
Repeat for the other two ailments.

File: Modules/CalcOffence-3_0.lua [hunk 5]
```diff
@@ -1645,6 +1662,30 @@ function calcs.offence(env, actor, activeSkill)
 		else
 			output.KnockbackChanceOnHit = skillModList:Sum("BASE", cfg, "EnemyKnockbackChance")
 		end
+		if skillModList:Sum("BASE", cfg, "ScorchChance") > 0 then
```
Comment (ppoelzl) [inline Modules/CalcOffence-3_0.lua:1665]: There are formatting issues with the indentation here.

File: Modules/CalcOffence-3_0.lua [hunk 6]
```diff
@@ -1599,6 +1602,20 @@ function calcs.offence(env, actor, activeSkill)
...
 		end
+		if skillModList:Flag(cfg, "CritAlwaysAltAilments") and not skillModList:Flag(cfg, "NeverCrit") and skillFlags.hit then
+			output.ScorchChanceOnCrit = 100
+			output.BrittleChanceOnCrit = 100
+			output.SapChanceOnCrit = 100
+		else
+			output.ScorchChanceOnCrit = 0
+			output.BrittleChanceOnCrit = 0
+			output.SapChanceOnCrit = 0
+			if skillModList:Flag(cfg, "CritAlwaysAltAilments") and not skillModList:Flag(cfg, "NeverCrit") then
```
Comment (ppoelzl) [inline Modules/CalcOffence-3_0.lua:1613]: This code block has to be like this, else the flags won't be correctly set on hitting skills:
```lua
if skillModList:Flag(cfg, "CritAlwaysAltAilments") and not skillModList:Flag(cfg, "NeverCrit") then
    skillFlags.inflictScorch = true
    skillFlags.inflictBrittle = true
    skillFlags.inflictSap = true
end
if skillModList:Flag(cfg, "CritAlwaysAltAilments") and not skillModList:Flag(cfg, "NeverCrit") and skillFlags.hit then
    output.ScorchChanceOnCrit = 100
    output.BrittleChanceOnCrit = 100
    output.SapChanceOnCrit = 100
else
    output.ScorchChanceOnCrit = 0
    output.BrittleChanceOnCrit = 0
    output.SapChanceOnCrit = 0
end
```
